___

<a href='http://www.pieriandata.com'> <img src='Pierian_Data_Logo.png' /></a>
___

# NumPy Operations

## Arithmetic

You can easily perform array with array arithmetic, or scalar with array arithmetic. Let's see some examples:

In [24]:
import numpy as np
arr = np.arange(0,10)

In [25]:
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [26]:
arr + arr

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [27]:
arr * arr

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

In [28]:
arr - arr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [29]:
# Warning on division by zero, but not an error!
# Just replaced with nan
arr/arr
#NumPy will issue a warning. it will say 'invalid value encountered in divide arr/arr'. it will give you a 'nan' and a warning instead of providing an error if you try to divide zero by zero. 

C:\Users\Surface Laptop\AppData\Local\Temp\ipykernel_14776\2012332712.py:3: RuntimeWarning: invalid value encountered in divide
  arr/arr


array([nan,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])

we also can perform array with scalar operations and what NumPy does is broadcasting that scalar (number) to every element in the array so the operation occurs also on an element by element basis.

In [30]:
arr +2

array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [31]:
# Also warning, but not an error instead infinity
1/arr

C:\Users\Surface Laptop\AppData\Local\Temp\ipykernel_14776\1360216608.py:2: RuntimeWarning: divide by zero encountered in divide
  1/arr


array([       inf, 1.        , 0.5       , 0.33333333, 0.25      ,
       0.2       , 0.16666667, 0.14285714, 0.125     , 0.11111111])

In [32]:
arr**3

array([  0,   1,   8,  27,  64, 125, 216, 343, 512, 729], dtype=int32)

## Universal Array Functions

Numpy comes with many [universal array functions](http://docs.scipy.org/doc/numpy/reference/ufuncs.html), which are essentially just mathematical operations you can use to perform the operation across the array. Let's show some common ones:

In [33]:
#Taking Square Roots
np.sqrt(arr)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ])

`np.sqrt(arr)` is **faster** and more efficient than `arr**(1/2)`. Here's why:

## **Performance Comparison**

1. **`np.sqrt(arr)`**
   - Uses **vectorized operations** optimized in C.
   - More efficient because NumPy directly calls **low-level, optimized math functions** (usually implemented in C or Fortran).
   - Avoids additional computational overhead.

2. **`arr**(1/2)`**
   - Uses Python's **exponentiation operator (`**`)**, which is **slower** than NumPy's built-in mathematical functions.
   - Internally calls `arr**0.5`, which involves **logarithms and exponentiation** (i.e., `x^y = exp(y * log(x))`), making it **less efficient**.

---

## **Benchmarking Example**
We can check the speed difference using `%timeit`:

In [34]:
array1 = np.random.rand(100000000)  

%timeit np.sqrt(array1)    # NumPy optimized sqrt function  
%timeit array1**(1/2)      # Exponentiation method  


410 ms ± 18.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
431 ms ± 38.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [35]:
#Calcualting exponential (e^)
np.exp(arr)

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03])

In [36]:
np.max(arr) #same as arr.max()

9

In [39]:
np.sin(arr)
#the output of np.sin() is in radians by default.
# Explanation:
    # In NumPy, trigonometric functions like np.sin(), np.cos(), and np.tan() assume the input is in radians.
    # If you want to use degrees instead, you must convert degrees to radians using np.radians().

array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
       -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849])

In [38]:
np.log(arr)

C:\Users\Surface Laptop\AppData\Local\Temp\ipykernel_14776\3120950136.py:1: RuntimeWarning: divide by zero encountered in log
  np.log(arr)


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458])

`np.log(arr)` calculates the **natural logarithm** (ln) of each element in the array `arr`.

## **Mathematical Definition**
\[
ln(x) = log_e(x)
\]
where \( e \) (Euler's number) is approximately **2.718**.

---

## **Example Usage**
```python
import numpy as np  

arr = np.array([1, np.e, 10])  
log_arr = np.log(arr)  

print(log_arr) #output: [0.         1.         2.30258509]


# Great Job!

That's all we need to know for now!